### DISCUSSION TITLE
Annotated objects with `init=False` gives "non-default argument  follows default argument" error

### DISCUSSION THREAD

**Discussion by mohammad-reza-rahbar** [sqlalchemy/sqlalchemy#12107] [Usage Questions] [answered]
When I use:

    date_created = Annotated[
        datetime,
        mapped_column(
            nullable=False,
            insert_default=lambda: datetime.now(timezone.utc),
            init=False,
        ),
    ]

as described [here](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapping-whole-column-declarations-to-python-types), I get the following error:

> sqlalchemy.exc.InvalidRequestError: Python dataclasses error encountered when creating dataclass for 'User': TypeError("non-default argument 'date_created' follows default argument 'is_active'"). Please refer to Python dataclasses documentation for additional information. (Background on this error at: https://sqlalche.me/e/20/dcte)

but when I directly use:

    date_created: Mapped[datetime] = mapped_column(
        nullable=False,
        insert_default=lambda: datetime.now(timezone.utc),
        init=False,
    )

the error disappears. Could you please tell me why?

**Comment by zzzeek** [comment, marked as answer]
~~this is a regression based on what we have defined as our behavior~~ this is now the current behavior, the docs need to be updated.  at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integration-with-annotated we define that while typing tools will not accommodate things like init=False wihtin Annotated, it should work at runtime.  it looks like  73a273c90cda2369ec071 in 2.0.14 broke this contract.


however, in 2.0.22 we added additional deprecation warnings and decided to have the ORM be fully consitent at runtime with what typing tools can provide so you will see this warning: ``SADeprecationWarning: Argument 'init' is a dataclass argument and cannot be specified within a mapped_column() bundled inside of an Annotated object``

**Reply by zzzeek** [reply]
this is now the expected behavior and the docs at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integration-with-annotated are inaccurate and im updating them now
